File: LeetCode/CPP/121.best-time-to-buy-and-sell-stock.cpp
/*
 * @lc app=leetcode id=121 lang=cpp
 *
 * [121] Best Time to Buy and Sell Stock
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        // By DP
        int max_price_yet = 0;
        if (prices.size() == 0)
        {
            return max_price_yet;
        }

        int min_ele_yet = prices[0];
        int len = prices.size();

        // keep finding the minimum element and calculate the max profit to ith day
        for (int i = 0; i < len; i++)
        {
            if (prices[i] < min_ele_yet)
            {
                min_ele_yet = prices[i];
            }
            max_price_yet = max(max_price_yet,
                                prices[i] - min_ele_yet);
        }
        return max_price_yet;
    }
};
// @lc code=end

File: LeetCode/Python/121.best-time-to-buy-and-sell-stock.py
#
# @lc app=leetcode id=121 lang=python3
#
# [121] Best Time to Buy and Sell Stock
#
from typing import List

# @lc code=start


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_price = 0
        curr_min = prices[0]
        for i in range(1, len(prices)):
            curr_min = min(prices[i - 1], curr_min)
            max_price = max(max_price, prices[i] - curr_min)

        return max_price


# @lc code=end
